README.md
# Fetch JSON Resource


[![Code Smells](https://sonarcloud.io/api/project_badges/measure?project=craftedsystems_fetch-json-resource&metric=code_smells)](https://sonarcloud.io/dashboard?id=craftedsystems_fetch-json-resource)
[![Coverage](https://sonarcloud.io/api/project_badges/measure?project=craftedsystems_fetch-json-resource&metric=coverage)](https://sonarcloud.io/dashboard?id=craftedsystems_fetch-json-resource)
[![Maintainability Rating](https://sonarcloud.io/api/project_badges/measure?project=craftedsystems_fetch-json-resource&metric=sqale_rating)](https://sonarcloud.io/dashboard?id=craftedsystems_fetch-json-resource)
[![Reliability Rating](https://sonarcloud.io/api/project_badges/measure?project=craftedsystems_fetch-json-resource&metric=reliability_rating)](https://sonarcloud.io/dashboard?id=craftedsystems_fetch-json-resource)
[![Security Rating](https://sonarcloud.io/api/project_badges/measure?project=craftedsystems_fetch-json-resource&metric=security_rating)](https://sonarcloud.io/dashboard?id=craftedsystems_fetch-json-resource)
[![SonarCloud Bugs](https://sonarcloud.io/api/project_badges/measure?project=craftedsystems_fetch-json-resource&metric=bugs)](https://sonarcloud.io/component_measures/metric/reliability_rating/list?id=craftedsystems_fetch-json-resource)
[![Sonarcloud Status](https://sonarcloud.io/api/project_badges/measure?project=craftedsystems_fetch-json-resource&metric=alert_status)](https://sonarcloud.io/dashboard?id=craftedsystems_fetch-json-resource) 
[![SonarCloud Vulnerabilities](https://sonarcloud.io/api/project_badges/measure?project=craftedsystems_fetch-json-resource&metric=vulnerabilities)](https://sonarcloud.io/component_measures/metric/security_rating/list?id=craftedsystems_fetch-json-resource)
[![Technical Debt](https://sonarcloud.io/api/project_badges/measure?project=craftedsystems_fetch-json-resource&metric=sqale_index)](https://sonarcloud.io/dashboard?id=craftedsystems_fetch-json-resource)

## Usage

```bash
# Install
npm i fetch-json-resource
```

```javascript
import fetchJsonResource from 'fetch-json-resource';

fetchJsonResource('https://jsonplaceholder.typicode.com/todos/1')
  .then((json) => console.log(json))
  .catch((error) => console.log(error));
```

## Dev

```bash
# Install
npm i

# Dev
npm run dev

# Build
npm run build
```


CHANGELOG.md
# Release Notes

## 0.1.0 - 2019-11-11

### Added

### Changed

### Fixed

### Security


test/test.js
import test from 'ava';

test.todo('will think about writing this later');


src/fetchResource.js
// @ts-check

/**
 * Map network error
 * @param {object} error
 */
function mapNetworkError(error) {
  if (error && error.code === 20 && error.name === 'AbortError') {
    // Abort controller event
    return {
      code: -1,
      id: 'network-timeout',
      message: 'Network timeout',
      type: 'app:network-timeout',
    }
  }
  return {
    code: -1,
    id: 'network-error',
    message: error.message,
    type: 'app:network-error',
  }
}

/**
 * Check for AbortController support
 * base on https://github.com/mo/abortcontroller-polyfill/blob/f20480d50d642d65790ee1af80e54a71506bcd46/src/abortcontroller-polyfill.js
 */
function hasAbortController() {
  const self = typeof window.self !== 'undefined'
    ? window.self
    : global;

  // Note that the "unfetch" minimal fetch polyfill defines fetch() without
  // defining window.Request, and this polyfill need to work on top of unfetch
  // so the below feature detection needs the !self.AbortController part.
  // The Request.prototype check is also needed because Safari versions 11.1.2
  // up to and including 12.1.x has a window.AbortController present but still
  // does NOT correctly implement abortable fetch:
  // https://bugs.webkit.org/show_bug.cgi?id=174980#c2
  // @ts-ignore
  // eslint-disable-next-line no-prototype-builtins
  const polyfillNeeded = (typeof (self.Request) === 'function' && !self.Request.prototype.hasOwnProperty('signal')) || !self.AbortController; // NOSONAR
  return !polyfillNeeded;
}

/**
 * Add AbortController
 * Use AbortController to be able to send abort signal to fetch and handle network timeouts.
 * @param {object} options
 */
function addAbortController(options) {
  const controller = new AbortController();
  // Abort request after 10 sec
  setTimeout(() => controller.abort(), 10000);
  return {
    ...options,
    signal: controller.signal,
  };
}

/**
 * Get last error type segment
 * @param {string} type
 */
function getLastErrorTypeSegment(type) {
  if (!type) return null;
  return type.split(':').pop();
}

/**
 * Parse response as JSON
 * @param {object} response
 */
async function toJSON(response) {
  try {
    const json = await response.json();
    if (json.code && typeof json.id === 'undefined') {
      json.id = getLastErrorTypeSegment(json.type);
    }
    return json;
  } catch(err) {
    return {
      code: -1,
      id: 'unknown',
      message: err.message,
      type: 'app:unknown',
    };
  }
}

/**
 * Fetch resource
 * @param {string} url
 * @param {object} options
 */
export default async function fetchResource(url, options) {
  const fetchOptions = hasAbortController()
    ? addAbortController(options)
    : options;
  return fetch(url, fetchOptions)
    .then((response) => toJSON(response))
    .catch((error) => mapNetworkError(error));
}


src/index.mjs
// @ts-check

/**
 * Map network error
 * @param {object} error
 */
function mapNetworkError(error) {
  if (error && error.code === 20 && error.name === 'AbortError') {
    // Abort controller event
    return {
      code: -1,
      id: 'network-timeout',
      message: 'Network timeout',
      type: 'app:network-timeout',
    }
  }
  return {
    code: -1,
    id: 'network-error',
    message: error.message,
    type: 'app:network-error',
  }
}

/**
 * Check for AbortController support
 * base on https://github.com/mo/abortcontroller-polyfill/blob/f20480d50d642d65790ee1af80e54a71506bcd46/src/abortcontroller-polyfill.js
 */
function hasAbortController() {
  const self = window && typeof window.self !== 'undefined'
    ? window.self
    : global;

  // Note that the "unfetch" minimal fetch polyfill defines fetch() without
  // defining window.Request, and this polyfill need to work on top of unfetch
  // so the below feature detection needs the !self.AbortController part.
  // The Request.prototype check is also needed because Safari versions 11.1.2
  // up to and including 12.1.x has a window.AbortController present but still
  // does NOT correctly implement abortable fetch:
  // https://bugs.webkit.org/show_bug.cgi?id=174980#c2
  // @ts-ignore
  // eslint-disable-next-line no-prototype-builtins
  const polyfillNeeded = (typeof (self.Request) === 'function' && !self.Request.prototype.hasOwnProperty('signal')) || !self.AbortController; // NOSONAR
  return !polyfillNeeded;
}

/**
 * Add AbortController
 * Use AbortController to be able to send abort signal to fetch and handle network timeouts.
 * @param {object} options
 */
function addAbortController(options) {
  const controller = new AbortController();
  // Abort request after 10 sec
  setTimeout(() => controller.abort(), 10000);
  return {
    ...options,
    signal: controller.signal,
  };
}

/**
 * Get last error type segment
 * @param {string} type
 */
function getLastErrorTypeSegment(type) {
  if (!type) return null;
  return type.split(':').pop();
}

/**
 * Parse response as JSON
 * @param {object} response
 */
async function toJSON(response) {
  try {
    const json = await response.json();
    if (json.code && typeof json.id === 'undefined') {
      json.id = getLastErrorTypeSegment(json.type);
    }
    return json;
  } catch(err) {
    return {
      code: -1,
      id: 'unknown',
      message: err.message,
      type: 'app:unknown',
    };
  }
}

/**
 * Fetch resource
 * @param {string} url
 * @param {object} options
 */
export default async function fetchResource(url, options) {
  const fetchOptions = hasAbortController()
    ? addAbortController(options)
    : options;
  return fetch(url, fetchOptions)
    .then((response) => toJSON(response))
    .catch((error) => mapNetworkError(error));
}
